Adding demo notebook and clean up

lastmansleeping opened the pull request:
This PR covers the following changes:

- activate notebook
- saving model layers separately for transfer learning
- removing weights and biases from ml4ir
- general cleanup

lastmansleeping commented on python/ml4ir/base/io/local_io.py:
```diff
@@ -251,3 +252,39 @@ def rm_file(self, file_path: str):
             os.remove(file_path)
             self.log("File deleted : {}".format(file_path))
+
+    def save_numpy_array(self, np_array, file_path: str, allow_pickle=True, zip=True, **kwargs):
```
@jakemannix I had to do this because there is an issue with saving a list of numpy arrays directly and tensorflow layer weights are lists of arrays unfortunately. `savez` internally saves each list item in separate files and abstracts this out for us.

lastmansleeping replied:
Added.

lastmansleeping commented on python/ml4ir/base/model/callbacks/debugging.py:
```diff
@@ -0,0 +1,52 @@
+from tensorflow.keras import callbacks
+
+
+class DebuggingCallback(callbacks.Callback):
```
Moving this out because `mypy` wouldn't let me commit due to code complexity.

lastmansleeping commented on python/ml4ir/base/model/optimizer.py:
```diff
@@ -7,17 +7,20 @@
-    learning_rate_decay: float,
-    learning_rate_decay_steps: int,
-    gradient_clip_value: float,
+    learning_rate_decay: float = 1.0,
```
Setting default values here.

Ullimague replied:
One of the drawback here is that we are overriding default tf values (with the same values, which makes things not robust/ideal).

lastmansleeping replied:
Yeah, I agree. Thoughts on how we should go about this? If I don't specify defaults, then it will expect us to pass this argument to the `get_optimizer` function each time. I can wrap it in if-else clauses, but it will be very dirty.

balikasg replied:
The code here looks for different optimizers based on keys that are configured on `ml4ir.base.config.keys`. We can come up with a way that skips this, and given the users choice directly looks at tensorflow.keras.optimizers.USER_CHOICE, if exists it returns it, if not it returns SGD with a message that it was not found (or simply fails with the message). This removes the need to maintain `ml4ir.base.config.keys`, fizes the problem of default arguments as they will be maintained. Arguments can be passed as done here, all being optional apart from the optimizer_key. Further args, spacific to each optimizer, can be passed with the (ugly) **kwards. The last advantage is that we expose the full functionality of the optimizers to a user.

lastmansleeping replied:
@balikasg I'm not sure I know how to do this -> "looks at tensorflow.keras.optimizers.USER_CHOICE, if exists it returns it". I need an enum because when using ml4ir as a script people will enter a string. I need to a way to map that string to a subclass of `tensorflow.keras.optimizers`

lastmansleeping commented on python/ml4ir/base/model/relevance_model.py:
```diff
@@ -234,16 +246,25 @@ def fit(
+                else:
+                    train_metrics[metric] = value[0]
+
+            return train_metrics
```
Returning metrics to be stored for experiment tracking. Since we are not using weights and biases, I am creating a dict manually.

Ullimague replied:
This info needs to be part of python doc. Ie explicitly define what we are returning.

lastmansleeping replied:
Added.

lastmansleeping commented on python/ml4ir/base/model/relevance_model.py:
```diff
@@ -381,6 +404,16 @@ def save(
         )
+
+        # Save individual layer weights
+        self.file_io.make_directory(os.path.join(model_file, "layers"), clear_dir=True)
```
Saving weights of each layer so that we can load it later in a new model for transfer learning.

Ullimague replied:
Can you add this info as pydoc?

lastmansleeping replied:
Added.

lastmansleeping commented on python/ml4ir/base/pipeline.py:
```diff
@@ -182,54 +179,6 @@ def validate_args(self):
 
         return self
 
-    def setup_experiment_tracking_config(self):
```
Not needed anymore. This is wandb related.

Ullimague commented on python/ml4ir/base/model/optimizer.py:
```diff
@@ -7,17 +7,20 @@
-        decay_rate=learning_rate_decay,
-        staircase=True,
-    )
+    if learning_rate_decay <= 1.0:
```
Why do we decide on this? We also need to add a doc for this.

balikasg replied:
Needs to be less than one to be decay. I agree with all your comments on doc.

Ullimague commented on python/ml4ir/base/model/relevance_model.py:
```diff
@@ -234,16 +246,25 @@ def fit(
+            # Write metrics for experiment tracking
+            train_metrics = dict()
+            for metric, value in history.history.items():
+                if not metric.startswith("val_"):
```
Seems weird, can't we have `if metric.startswith("train_")`?

lastmansleeping replied:
This is not a check though. It's the key for the dictionary. How can I do `.startswith`?

Ullimague commented on python/ml4ir/base/pipeline.py:
```diff
@@ -351,17 +301,62 @@ def run(self):
+        self.finish(job_status, job_info)
+
+    def finish(self, job_status, job_info):
+        """
```
We need to add pydoc about what finish does.

lastmansleeping replied:
Added.

Ullimague commented on python/ml4ir/base/model/relevance_model.py:
```diff
@@ -325,20 +346,22 @@ def evaluate(
         Args:
```
line 323: "Evaluate the ranking model" it's actually the base model. And we would need to provide more info, because we do more than just evaluating it (returning metrics, saving layers, etc.)

Ullimague commented on python/ml4ir/base/model/relevance_model.py:
```diff
@@ -325,20 +346,22 @@ def evaluate(
 
         Returns:
-            metrics and groupwise metrics as pandas DataFrames
+            metrics and groupwise metrics as pandas DataFrames and flattened metrics as a dictionary
```
We should define here what is expected as output. Reading the code we just see: `returns None, None, dict`
Ideal we should add an example. And what of each value in the Triple means.

lastmansleeping replied:
I've added what each value in the triple means, right? It is (metrics, groupwise_metrics, metrics_dict)
I can add example.

lastmansleeping replied:
Added.

Ullimague commented on python/ml4ir/base/model/scoring/scoring_model.py:
```diff
@@ -72,11 +72,11 @@ def final_activation_op(self, scores, metadata_features):
 
 class RelevanceScorer(ScorerBase):
```
We need to add pydoc.

Ullimague commented on python/ml4ir/base/model/scoring/scoring_model.py:
```diff
@@ -72,11 +72,11 @@ def final_activation_op(self, scores, metadata_features):
 
 class RelevanceScorer(ScorerBase):
     def architecture_op(self, train_features, metadata_features):
```
Same here: need to add pydoc. What is an architecture_op?

lastmansleeping replied:
I have a work item to add pydocs for all methods. Will do it.

Ullimague reviewed (commented):
Early comment to help iterate. Waiting for the tests for continuing the review :)

lastmansleeping commented on python/ml4ir/applications/classification/tests/test_classification_model.py:
```diff
@@ -17,22 +17,18 @@ def run_default_pipeline(self, data_format: str):
         # Fix random seed values for repeatability
         self.set_seeds()
 
-        args: Namespace = self.get_overridden_args(data_format)
```
I will add it. It was giving me errors while committing because you had not imported Namespace in this module. I think you use an IDE and it doesn't catch these errors because imports from all modules are in same context.

lastmansleeping commented on python/ml4ir/base/features/feature_config.py:
```diff
@@ -336,26 +369,29 @@ class SequenceExampleFeatureConfig(FeatureConfig):
     """Feature config overrides for data containing SequenceExample protos"""
 
     def __init__(self, features_dict, logger):
+        super(SequenceExampleFeatureConfig, self).__init__(features_dict, logger)
```
There's a command line argument called `tfrecord_type` which can be set to `sequence_example` or `example`

balikasg commented on python/ml4ir/base/model/scoring/interaction_model.py:
```diff
@@ -39,8 +39,8 @@ class UnivariateInteractionModel(InteractionModel):
         self,
         feature_config: FeatureConfig,
-        feature_layer_keys_to_fns: dict,
         tfrecord_type: str,
```
A naive ask here: what is exactly an Interaction model? why  do we call it like this? Is it a family of feedforward NNs or something in these lines? I have seen it for ranking and ep but I do not grasp the term well

lastmansleeping replied:
We can work on the term. But it is the first layer of the model where you define the feature functions for each feature and then combine them into a dense layer.

Use this layer to define embeddings, LSTMs, etc.

lastmansleeping commented on python/ml4ir/applications/ranking/tests/test_transfer_learning.py:
```diff
@@ -0,0 +1,130 @@
+from ml4ir.base.features.feature_config import FeatureConfig
+
+
+class RankingTransferLearningTest(RankingTestBase):
```
@Ullimague Added the tests here. I tried to cover all scenarios.

lastmansleeping commented on python/ml4ir/base/features/feature_config.py:
```diff
@@ -102,29 +108,28 @@ def get_instance(feature_config_dict: dict, tfrecord_type: str, logger: Logger):
 
-    def extract_features(self, features_dict, logger: Optional[Logger] = None):
-
+    def extract_features(self):
```
Added.

lastmansleeping commented:
@Ullimague I've added the unit test, pydocs and addressed all your review comments. Let me know if this works. I have a parallel work item (WIP) to add and consolidate the pydocs across ml4ir.

Ullimague reviewed (approved):
Thanks for the follow-ups @lastmansleeping!
